### DISCUSSION TITLE
No interrupt safety on GPIO pins

### DISCUSSION THREAD

**Discussion by hellebauer** [bxparks/AceWire#10] [Q&A]
Using native Wire.h causes occasional wedging in my app (Arduino IDE, MEGA2560) , thus I'm looking at AceWire.

In the docs it says about SimpleWireInterface and SimpleWireFastInterface : 

>....... These 2 implementations can be sufficient in many simple projects that can deal with the following limitations:   no interrupt safety on GPIO pins   ..........

What exactly do you mean by "no interrupt safety on GPIO pins" ?
My app uses PCINT9 (RX3) on a MEGA2560, would it be safe to use  SimpleWireFastInterface or SimpleWireInterface  ?

Thanks a heap for a quick word of wisdom

**Comment by bxparks** [comment]
The SimpleWireInterface and SimpleWireFastInterface assume that they have complete control over the GPIO registers for the duration of the I2C communication. If your application is using interrupts, then it must make sure that the interrupt service routine (ISR) does not conflict with the GPIO registers.

One way is to make sure that the ISR never touches anything related to the GPIO pins used by SimpleWireInterface or SimpleWireFastInterface. Another way might be to disable interrupts just before `beginTransmission()` and reenable them after `endTransmission()`. I have never tested this unfortunately.

Also, my understanding is that the normal Arduino `digitalWrite()` and `digitalRead()` functions provide some level of interrupt safety. However, the `digitalWriteFast()` and `digitalReadFast()` manipulate the GPIO registers directly and provide absolutely no interrupt safety. So if you are using SimpleWireFastInterface, you have to be even more careful.

I don't have much experience with using interrupts, but I know enough to realize that interrupt safety is a tricky thing in non-trivial situations. I hope this helps.
